Assets/1SimpleNavigation/SimpleNavController.cs

```diff
@@ -1,47 +1,93 @@
-
-            if(a.MaxStep > 0 && a.StepCount >= a.MaxStep) { isEnd = true; reachedTarget = false; EndAll(); } // Max stepcount reached
-            if(Distance(a, new Vector3(0, 0, 0)) > boundsRadius) { isEnd = true; reachedTarget = false; EndAll(); } // Out of bounds
-            if(Distance(a, target.transform.localPosition) < 1.5f) { a.AddReward(1.0f); reachedTarget = true; isEnd = true; EndAll(); } // Reached target
+    private int step = 0;
+    private bool reachedTarget;
+    private const float TargetReachDistance = 1.5f;
+
+    protected override void EndCase()
+    {
+        bool isEnd = CheckEndConditions();
+        if (isEnd) EndAll();
+    }
+
+    private bool CheckEndConditions()
+    {
+        return agents.Any(agent =>
+        {
+            step = agent.StepCount;
+
+            if (agent.MaxStep > 0 && agent.StepCount >= agent.MaxStep)
+            {
+                reachedTarget = false;
+                return true;
+            }
+
+            if (IsOutOfBounds(agent))
+            {
+                reachedTarget = false;
+                return true;
+            }
+
+            if (HasReachedTarget(agent))
+            {
+                agent.AddReward(1.0f);
+                reachedTarget = true;
+                return true;
+            }
+
+            return false;
         });
     }
 
-    protected override void CalculateReward(ABMAgent agent) {
-        agent.AddReward(1f / (Distance(agent, target.transform.localPosition)+0.1f));
+    private bool IsOutOfBounds(ABMAgent agent)
+    {
+        return Distance(agent, Vector3.zero) > boundsRadius;
     }
 
-    // Distance of Agent (a) to target (t)
-    private float Distance(ABMAgent a, Vector3 t) {
-        return Vector3.Distance(a.transform.localPosition, t);
+    private bool HasReachedTarget(ABMAgent agent)
+    {
+        return Distance(agent, target.transform.localPosition) < TargetReachDistance;
     }
 
-    protected override void SetupEpisode() {
+    protected override void CalculateReward(ABMAgent agent)
+    {
+        agent.AddReward(1f / (Distance(agent, target.transform.localPosition) + 0.1f));
+    }
+
+    private float Distance(ABMAgent agent, Vector3 targetPosition)
+    {
+        return Vector3.Distance(agent.transform.localPosition, targetPosition);
+    }
+
+    protected override void SetupEpisode()
+    {
         base.SetupEpisode();
-        target.transform.localPosition = new Vector3(RandomRadius(1.5f), RandomRadius(1.5f), RandomRadius(1.5f)); // Spawn Randomly
+        SpawnTarget();
+    }
+
+    private void SpawnTarget()
+    {
+        target.transform.localPosition = new Vector3(
+            RandomRadius(TargetReachDistance),
+            RandomRadius(TargetReachDistance),
+            RandomRadius(TargetReachDistance)
+        );
     }
 
-    public float RandomRadius(float sub) {
-        return UnityEngine.Random.Range(-boundsRadius+sub, boundsRadius-sub);
+    private float RandomRadius(float buffer)
```

Comment:
this one also came from Sonnet 3.5, I think it wasn't necessary to make it public